packages/nuka/src/hooks/use-transition.ts

```diff
@@ -0,0 +1,48 @@
+import { useEffect, useState, useRef } from 'react';
+import { D3EasingFunctions } from 'src/types';
+import * as d3Ease from 'victory-vendor/d3-ease';
+
+/**
+ * Provides an interpolated value from [0, 1] based on a provided duration
+ * and d3-ease animation timing function name.
+ */
+export const useTransition = (
+  duration: number, // in milliseconds
+  animationTimingFunction: D3EasingFunctions,
+  deps?: React.DependencyList
+): number => {
+  const fps = 1000 / 60;
+  const [normalizedTime, setNormalizedTime] = useState(0);
+  const startTime = useRef(Date.now());
+  const frameLoop = useRef<NodeJS.Timer | null>(null);
+
+  const clearFrameLoop = () => {
+    if (frameLoop.current) clearInterval(frameLoop.current);
+  };
+
+  const resetFrameLoop = () => {
+    startTime.current = Date.now();
+    setNormalizedTime(0);
+  };
+
+  const updateFrameLoop = () => {
+    frameLoop.current = setInterval(() => {
+      const currentTime = Date.now();
+      const normalizedTime = Math.min(
+        1,
+        (currentTime - startTime.current) / duration
+      );
+      setNormalizedTime(normalizedTime);
+      if (normalizedTime === 1) clearFrameLoop();
+    }, fps);
```

Comment:
This type of thing normally concerns me a bit, since we're setting React state _a lot_ here. However, I think we're doing a similar thing with out dragging – so this probably isn't out of norm for our existing setup.

For a side note, in terms of React performance, it's usually best to:

- use a `requestAnimationFrame` loop instead of `setInterval` and perform similar logic.
- use a DOM reference and imperatively update the DOM property, as opposed to using React state – so that we don't have to go through 60 React DOM reconciliations per second.
- Once an animation finishes, _then_ you set React state to get the rest of your React logic up to date with your imperative DOM manipulations.

However, that's generally much more complex and less readable than using React state, so maybe not something we worry about here.

Reply:
I could try to implement it this way in a follow up PR.

Reply:
RE: Grant's feedback: I did a quick check of [how popular animation libraries set up their APIs](https://blog.logrocket.com/best-react-animation-library-top-7-libraries-compared/) to update the DOM, and it appears using React state / values in JSX to update is pretty standard, so I would assume it's decently performant as long as there's not too much logic/UI stuffed in there.

I have heard the `requestAnimationFrame` advice before, too, and usually rely on that method for anything related to animation.